crypto/evp/evp_enc.c

```diff
@@ -1600,20 +1602,24 @@ int EVP_CIPHER_up_ref(EVP_CIPHER *cipher)
 {
     int ref = 0;
 
-    CRYPTO_UP_REF(&cipher->refcnt, &ref, cipher->lock);
+    if (cipher->origin == EVP_ORIG_DYNAMIC)
+        CRYPTO_UP_REF(&cipher->refcnt, &ref, cipher->lock);
     return 1;
 }
 
 void EVP_CIPHER_free(EVP_CIPHER *cipher)
 {
     int i;
 
-    if (cipher == NULL)
+    if (cipher == NULL || cipher->origin == EVP_ORIG_GLOBAL)
         return;
 
-    CRYPTO_DOWN_REF(&cipher->refcnt, &i, cipher->lock);
-    if (i > 0)
-        return;
+    if (cipher->origin == EVP_ORIG_DYNAMIC) {
+        CRYPTO_DOWN_REF(&cipher->refcnt, &i, cipher->lock);
+        if (i > 0)
+            return;
+    }
+
     ossl_provider_free(cipher->prov);
     CRYPTO_THREAD_lock_free(cipher->lock);
     OPENSSL_free(cipher);
```

Comment:
This suffers from the same problem as `EVP_MD_free()` did.  I suggest making the same kind of changes you did with `EVP_MD_free()` and `EVP_MD_meth_free()`

Reply:
Updated commit (with accurate comments now :) pushed.  Thanks.